feat: openapi implementation

VishnuSanal opened the pull request:
**Description**

This PR fixes #247

sansyrox commented:
Hey @VishnuSanal 👋 

Before you proceed further. Just one clarification OpenAPI is supposed to be generated *automatically* from the types.

VishnuSanal commented:
TODO:

- [ ] path param type
- [ ] path param required?:
- [ ] query params

sansyrox commented on robyn/__init__.py:
```diff
@@ -18,15 +20,16 @@
 
 
 config = Config()
+openapi = OpenAPI()
```
we don't need to make openapi global. use self.openapi to access it

sansyrox commented on robyn/__init__.py:
```diff
@@ -18,15 +20,16 @@
 from robyn.router import MiddlewareRouter, MiddlewareType, Router, WebSocketRouter
 from robyn.types import Directory
 from robyn.ws import WebSocket
+from .openapi import OpenAPI
```
could you do absolute imports e.g. `robyn.openapi`

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,206 @@
+        return json.dumps(self.openapi_schema)
+
+
+openapi = OpenAPI()
```
why export this?

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,206 @@
+from robyn import serve_html
+
+
+class OpenAPI:
```
also please add docstrings in the methods

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,206 @@
+import os
+from typing import List, Dict, Union, Any
+
+from robyn import serve_html
```
please import from source. this can introduce cyclic imports

sansyrox reviewed (commented):
Hey @VishnuSanal 👋 

I have a few suggestions that can improve the code style. These are just initial ones. I will do a thorough review when you put the PR up for review 😄

sansyrox commented:
Hey @VishnuSanal 👋 

Could you please summarise the PR in the description?

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,267 @@
+
+
+class OpenAPI:
+    def __init__(
```
please add docstrings here.

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,267 @@
+            component_path_items=openapi_component_path_items,
+        )
+
+    def build_schema(
```
please add docstrings here.

VishnuSanal replied:
what? isn't it already added for this function?

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,267 @@
+import robyn
+
+
+class OpenAPI:
```
and maybe we can initialise it as a dataclass. to make the interface look prettier.

sansyrox replied:
https://github.com/sparckles/Robyn/blob/6f76c64db2425ace6060a663bbd199e2f9373385/robyn/openapi_middleware.py

Have a look at this PR . This implementation looks much cleaner. We should include this in this PR

tavallaie replied:
we can use pydantic here instead of dataclass so we can slowly move to pydantic.

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,267 @@
+
+        :return: the "path" openapi object according to spec
+        """
+        modified_endpoint = endpoint
```
how is it modified here?

VishnuSanal replied:
Robyn uses `:param` for path params & swagger expects `{param}` -- this is being done here.

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,267 @@
+        if len(path_params) > 1:
+            path = path_params[0]
+
+            modified_endpoint = path[:-1] if path.endswith("/") else path
```
what is `path[:-1]` ? I'd recommend a bit more description here 😅

VishnuSanal replied:
Sure, that's just trimming out the trailing `/`

VishnuSanal replied:
try uring `removeSuffix` or `rstrip`

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,267 @@
+        get the swagger html page
+        @return: the swagger html page
+        """
+        json.dumps(self.openapi_schema)
```
not needed imo

VishnuSanal replied:
It is needed! The openapi.json file won't get generated automatically -- the swagger page will raise an error then.

VishnuSanal replied:
mention it in docstring.

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,267 @@
+        @return: the swagger html page
+        """
+        json.dumps(self.openapi_schema)
+        html_file = os.path.join(os.getcwd(), "robyn/swagger.html")
```
we should can use `pathlib` here https://docs.python.org/3/library/pathlib.html

sansyrox commented on testing_application/app.py:
```diff
@@ -0,0 +1,82 @@
+        openapi_contact_email="support@example.com",
+        openapi_license_name="Apache 2.0",
+        openapi_license_url="https://www.apache.org/licenses/LICENSE-2.0.html",
+        # openapi_servers=[
+        #     {"url": "/", "description": "Debug environment"},
+        #     {"url": "https://example.com/api/v1/", "description": "Production environment"},
+        # ],
+        # openapi_external_docs={
+        #     "description": "Find more info here",
+        #     "url": "https://example.com/"
+        # }
```
is this needed?

VishnuSanal replied:
These are another specs that the user may want to add. We're just supporting them.

sansyrox commented on testing_application/app.py:
```diff
@@ -0,0 +1,82 @@
+from typing import TypedDict
```
also one question could this be merged with the baseRoutes?

VishnuSanal commented:
note to self: check sanic's impl of openapi

tavallaie commented on testing_application/app.py:
```diff
@@ -0,0 +1,82 @@
+app.include_router(doctor_router)
+
+if __name__ == "__main__":
+    app.start()
```
I am just thinking that maybe we can add some unit test too?
I think we cannot test this feature with github actions

sansyrox commented on robyn/__init__.py:
```diff
@@ -1,11 +1,13 @@
 import asyncio
```
@VishnuSanal , one more suggestion. we would also need documentation around it.

VishnuSanal replied:
didn't get you? that's not a line I changed!

sansyrox replied:
@VishnuSanal , I mean that the entire implementation requires a new page in the docs section!

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,188 @@
+
+
+@dataclass
+class OpenAPIInfo:
```
@VishnuSanal please add docstrings to every data class

sansyrox commented on integration_tests/openapi/__init__.py:
```diff
@@ -0,0 +1,92 @@
+from typing import TypedDict
```
@VishnuSanal , we should add integration tests in the PR

tavallaie replied:
for each functions you need to make a test function for example:
```
def test_docs_handler(openapi_instance):
    html_response = openapi_instance.docs_handler()
    assert isinstance(html_response, str)
    assert "swagger.html" in html_response

def test_json_handler(openapi_instance):
    json_response = openapi_instance.json_handler()
    assert isinstance(json_response, str)
    assert "openapi" in json_response
    assert "info" in json_response
    assert "paths" in json_response
```

tavallaie replied:
another example:
@pytest.fixture
def openapi_instance():
    return OpenAPI()

def test_openapi_initialization(openapi_instance):
    spec = openapi_instance.openapi_spec
    assert spec["openapi"] == "3.0.0"
    assert spec["info"]["title"] == "Robyn API"
    assert spec["info"]["version"] == "1.0.0"
    assert "paths" in spec
    assert "components" in spec
    assert "servers" in spec
    assert "externalDocs" in spec

VishnuSanal replied:
I have added the tests for all functions -- PTAL. :) btw, thanks for the code!!

sansyrox commented on integration_tests/openapi/__init__.py:
why do we need to create a new module?

I'd recommend two ways to test it.

- First test it as a part of the main router in `integration_tests/base_routes.py` 
- Then test it as a part of a subrouter

tavallaie commented:
@sansyrox is there anything left for merging this PR? 
also, @VishnuSanal we can make example directory and put your app there alongside other example we will make.

sansyrox commented on docs_src/src/pages/documentation/example_app/openapi.mdx:
````diff
@@ -0,0 +1,123 @@
+                    description="Debug environment"
+                ),
+                Server(
+                    url="https://example.com/api/v1/",
+                    description="Production environment"
+                ),
+            ],
+            externalDocs=ExternalDocumentation(description="Find more info here", url="https://example.com/"),
+            components=Components(),
+        ),
+    ),
+)
+
+
+@app.get("/")
+async def welcome():
+    """welcome endpoint"""
+    return "hi"
+
+
+class GetRequestParams(TypedDict):
+    appointment_id: str
+    year: int
+
+
+@app.get("/api/v1/:name", openapi_tags=["Name"])
+async def get(r: Request, query_params=GetRequestParams):
+    """Get Name by ID"""
+    return jsonify(r.path_params)
+
+
+@app.delete("/users/:name", openapi_tags=["Name"])
+async def delete(r: Request):
+    """Delete Name by ID"""
+    return jsonify(r.path_params)
+
+
+if __name__ == "__main__":
+    app.start()
+```
````
Hey @VishnuSanal , please use `CodeGroup` like used in the other docs. 

Also, please provide - `typed` and an `untyped` version in the docs.

sansyrox commented on integration_tests/test_openapi.py:
```diff
@@ -0,0 +1,71 @@
+    return OpenAPI()
+
+
+def test_openapi_initialization(openapi_instance):
+    spec = openapi_instance.openapi_spec
+    assert spec["openapi"] == "3.0.0"
+    assert spec["info"]["title"] == "Robyn API"
+    assert spec["info"]["version"] == "1.0.0"
+    assert "paths" in spec
+    assert "components" in spec
+    assert "servers" in spec
+    assert "externalDocs" in spec
+
+
+def test_docs_handler(openapi_instance):
+    html_response = openapi_instance.docs_handler()
+    assert isinstance(html_response, FileResponse)
+
+
+def test_json_handler(openapi_instance):
+    json_response = openapi_instance.json_handler()
+    assert isinstance(json_response, str)
+    assert "openapi" in json_response
+    assert "info" in json_response
+    assert "paths" in json_response
+    assert "components" in json_response
+    assert "servers" in json_response
+    assert "externalDocs" in json_response
+
+
+def test_add_openapi_path(openapi_instance):
```
These are all unit tests. I don't like to keep unit tests in the repo,and I don't think these would be any different. The tests should be written in an integrated test format like done for all the other routes.

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,263 @@
+
+        :return: the "path" openapi object according to spec
+        """
+        # for converting path param from `:param` to `{param}`
```
could you write an example here -

e.g. robyn has paths like `/:url/:etc` whereas openapi requires path like `/{url}/{path}` ?

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,263 @@
+
+    def __post_init__(self):
+        """
+        initializes the openapi_spec dict
```
```suggestion
        Initializes the openapi_spec dict
```

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,263 @@
+
+    def add_openapi_path_obj(self, route_type: str, endpoint: str, openapi_summary: str, openapi_tags: list, signature: Signature = None):
+        """
+        adds the given path to openapi spec
```
```suggestion
        Adds the given path to openapi spec
```

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,263 @@
+
+    def add_subrouter_paths(self, subrouter_openapi):
+        """
+        adds the subrouter paths to main router's openapi specs
```
```suggestion
        Adds the subrouter paths to main router's openapi specs
```

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,263 @@
+
+    def get_path_obj(self, endpoint: str, summary: str, tags: list, query_params: TypedDict = None):
+        """
+        get the "path" openapi object according to spec
```
```suggestion
        Get the "path" openapi object according to spec
```

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,263 @@
+        side effect: this function also dumps the openapi json file
+        @return: the swagger html page
+        """
+        json.dumps(self.openapi_spec)
```
I believe we should check if there is an openapi file already present in that location. And json.dumps should be a part of a different function, which is being called in the `docs_handler` function.

VishnuSanal replied:
> I believe we should check if there is an openapi file already present in that location. 

this function only gets called whilst running the application. and, IMO the file should be overwritten since the app & the endpoints might have changed.

> And json.dumps should be a part of a different function, which is being called in the docs_handler function.

sure, done.

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,263 @@
+
+    def docs_handler(self):
+        """
+        get the swagger html page
```
Also, this description is not accurate for the docs_handler. 
> get the swagger HTML page

Get where? Why is it named a hander inside the OpenAPI class?

VishnuSanal replied:
i have rephrased the docstrings anyways

sansyrox commented on testing_application/app.py:
```diff
@@ -0,0 +1,89 @@
+                url="https://www.apache.org/licenses/LICENSE-2.0.html",
+            ),
+            servers=[
+                # Server(
+                #     url="//",
+                #     description="Debug environment"
+                # ),
+                # Server(
+                #     url="https://example.com/api/v1/",
+                #     description="Production environment"
+                # ),
```
Do we need this?

VishnuSanal replied:
need? no, but I just demonstrated an example the user can use. anyways, that file is to be removed later, as specified in the comment. I will remove it for now.

sansyrox commented on testing_application/app.py:
```diff
@@ -0,0 +1,89 @@
+@doctor_subrouter.delete("/users/:name/:age")
+async def delete_doctor(r: Request):
+    """Delete Doctor by ID"""
+    return jsonify(r.path_params)
```
we can just return the dictionary here. We don't need to use `jsonify`

sansyrox commented on testing_application/app.py:
do we even need this file?

VishnuSanal replied:
nops! I just demonstrated an example the user can use. anyways, that file is to be removed later, as specified in the comment. I will remove it for now.

sansyrox commented:
Hey @VishnuSanal ,

Still needs a few changes. Also, the CI is failing. Could you have a look?

sansyrox commented on docs_src/src/components/documentation/Guides.jsx:
```diff
@@ -34,6 +34,12 @@ const guides = [
+    href: '/documentation/example_app/openapi',
+    name: 'OpenAPI Specification',
+    description:
+      'Learn how to generate OpenAPI docs for your applications.',
```
```suggestion
      'Learn how OpenAPI docs are generate for your applications.',
```

sansyrox commented on docs_src/src/components/documentation/Guides.jsx:
```diff
@@ -34,6 +34,12 @@ const guides = [
       'Learn how to deploy your app to production and manage your deployments.',
   },
+  {
+    href: '/documentation/example_app/openapi',
```
@VishnuSanal , this is a core feature. We should add it in the `api_reference` as well 😄

sansyrox commented on docs_src/src/pages/documentation/example_app/openapi.mdx:
````diff
@@ -0,0 +1,223 @@
+<CodeGroup title="Basic App">
+
+```python {{ title: 'untyped' }}
+from typing import TypedDict
+
+from robyn import Robyn, Request, jsonify, OpenAPI
+from robyn.openapi import OpenAPIInfo, Contact, License, ExternalDocumentation, Components
````
@VishnuSanal , this is untyped 😅

sansyrox commented on docs_src/src/pages/documentation/example_app/openapi.mdx:
```diff
@@ -0,0 +1,223 @@
+            title="Sample App",
+            description=" This is a sample server application.",
+            termsOfService=" https://example.com/terms/",
+            version=" 1.2.2",
```
```suggestion
            version="1.2.2",
```

sansyrox commented on docs_src/src/pages/documentation/example_app/openapi.mdx:
```diff
@@ -0,0 +1,223 @@
+@app.get("/api/v1/:name", openapi_tags=["Name"])
+async def get(r: Request, query_params=GetRequestParams):
+    """Get Name by ID"""
+    return jsonify(r.path_params)
```
we don't need `jsonify`

sansyrox commented on docs_src/src/pages/documentation/example_app/openapi.mdx:
```diff
@@ -0,0 +1,223 @@
+            ),
+            servers=[
+                Server(
+                    url="//",
```
could you explain this?

sansyrox commented on docs_src/src/pages/documentation/example_app/openapi.mdx:
```diff
@@ -0,0 +1,223 @@
+@app.get("/api/v1/:name", openapi_tags=["Name"])
+async def get(r: Request, query_params=GetRequestParams):
+    """Get Name by ID"""
+    return jsonify(r.path_params)
```
jsonify!

sansyrox commented on robyn/__init__.py:
```diff
@@ -287,110 +295,174 @@ def inner(handler):
         """
         The @app.patch decorator to add a route with PATCH method
 
         :param endpoint [str]: [endpoint to server the route]
```
please update the docstring here

sansyrox commented on robyn/__init__.py:
```diff
@@ -287,110 +295,174 @@ def inner(handler):
+        openapi_tags: list = ["default"],
+    ):
         """
         The @app.delete decorator to add a route with DELETE method
```
docstring!

sansyrox commented on robyn/__init__.py:
```diff
@@ -287,110 +295,174 @@ def inner(handler):
         """
         The @app.put decorator to add a get route with PUT method
 
         :param endpoint str: endpoint to server the route
```
docstring

sansyrox commented on robyn/__init__.py:
```diff
@@ -287,110 +295,174 @@ def inner(handler):
         return inner
 
-    def get(self, endpoint: str, const: bool = False, auth_required: bool = False):
+    def get(
```
docstrings!

sansyrox commented on robyn/__init__.py:
```diff
@@ -287,110 +295,174 @@ def inner(handler):
         return inner
 
-    def post(self, endpoint: str, auth_required: bool = False):
+    def post(
```
docstring!

sansyrox commented on robyn/__init__.py:
```diff
@@ -287,110 +295,174 @@ def inner(handler):
         return inner
 
-    def head(self, endpoint: str, auth_required: bool = False):
+    def head(
```
docstring!

sansyrox commented on robyn/__init__.py:
```diff
@@ -287,110 +295,174 @@ def inner(handler):
+        self,
+        endpoint: str,
+        auth_required: bool = False,
+        openapi_tags: list = ["default"],
```
docstring!

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,271 @@
+import json
+import re
+from dataclasses import dataclass, field, asdict
+from inspect import Signature
+from pathlib import Path
+from typing import List, Dict, TypedDict, Optional
```
could you use `isort`(https://pycqa.github.io/isort/) here?

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,271 @@
+    The contact information for the exposed API.
+    (https://swagger.io/specification/#contact-object)
+
+    @param name The identifying name of the contact person/organization.
+    @param url The URL pointing to the contact information. This MUST be in the form of a URL.
+    @param email The email address of the contact person/organization. This MUST be in the form of an email address.
```
@VishnuSanal , the docstring should also have a type associated to it.
```suggestion
    @param name: Optional[str] The identifying name of the contact person/organization.
    @param url: Optional[str] The URL pointing to the contact information. This MUST be in the form of a URL.
    @param email: Optional[str] The email address of the contact person/organization. This MUST be in the form of an email address.
```

sansyrox commented on robyn/openapi.py:
all the docstrings require a param of string here.

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,271 @@
+
+        :return: the "path" openapi object according to spec
+        """
+        # robyn has paths like /:url/:etc whereas openapi requires path like /{url}/{path}
+        # this function is used for converting path params to the required form
+        # initialized with endpoint for handling endpoints without path params
```
nicee

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,271 @@
+                )
+                endpoint_with_path_params_wrapped_in_braces += "/{" + param_name + "}"
+
+        if query_params:
+            for query_param in query_params.__annotations__:
+                # ugly hack!
+                # returns "<class 'int'>" -- this line strips out the type a.k.a int from it
+                param_type = re.findall("'[a-z]+'", str(query_params.__annotations__[query_param]))[0].replace("'", "")
```
is there a better way for this?

VishnuSanal replied:
hey @sansyrox I can't figure out a way to do it, can you give some help & point me in the right direction?!

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,271 @@
+            },
+        }
+
+    def dump_openapi_spec_file(self):
```
could you add return types here? in this file

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,271 @@
+        path_params = endpoint.split(":")
+        openapi_parameter_object = []
+
+        if len(path_params) > 1:
```
I feel we can write it a bit better.

VishnuSanal replied:
I have updated the implementation, and used better variable names, please check. :)

VishnuSanal commented:
- [ ] TODO: fix CI

sansyrox commented on docs_src/src/components/documentation/ApiDocs.jsx:
```diff
@@ -106,6 +106,12 @@ const guides = [
     description: 'Learn about GraphQL Support in Robyn.',
   },
+  {
+    href: '/documentation/example_app/openapi',
```
Could you copy this the api reference folder too? And use that path

sansyrox commented on docs_src/src/pages/documentation/example_app/openapi.mdx:
```diff
@@ -0,0 +1,203 @@
+                email="support@example.com",
+            ),
+            license=License(
+                name="GPLv3.0+",
```
Used bsd2.0 license please

sansyrox commented on docs_src/src/pages/documentation/example_app/openapi.mdx:
```diff
@@ -0,0 +1,203 @@
+@app.get("/api/v1/:name", openapi_tags=["Name"])
+async def get(r, query_params=GetRequestParams):
+    """Get Name by ID"""
+    return r.path_params
```
This should be query params, no?

sansyrox commented on docs_src/src/pages/documentation/example_app/openapi.mdx:
```diff
@@ -0,0 +1,203 @@
+
+</CodeGroup>
+
+With the reference documentation deployed and running smoothly, Batman had a powerful new tool at his disposal. The Robyn framework had provided him with the flexibility, scalability, and performance needed to create an effective crime-fighting application, giving him a technological edge in his ongoing battle to protect Gotham City.
```
Nice

sansyrox commented on robyn/__init__.py:
```diff
@@ -222,6 +226,9 @@ def start(self, host: str = "127.0.0.1", port: int = 8080, _check_port: bool = T
         """
 
+        self.add_route(route_type=HttpMethod.GET, endpoint="/openapi.json", handler=self.openapi.json_handler, is_const=False)
+        self.add_route(route_type=HttpMethod.GET, endpoint="/docs", handler=self.openapi.docs_handler, is_const=False)
```
These should be const requests

sansyrox commented on docs_src/src/components/documentation/Navigation.jsx:
```diff
@@ -293,6 +297,10 @@ export const navigation = [
       },
+      {
+        title: 'OpenAPI Specification',
+        href: '/documentation/example_app/openapi',
```
should be api_reference here

sansyrox commented on docs_src/src/components/documentation/Navigation.jsx:
```diff
@@ -293,6 +297,10 @@ export const navigation = [
         title: 'Advanced Features',
       },
+      {
+        title: 'OpenAPI Specification',
```
```suggestion
        title: 'OpenAPI Documentation',
```

sansyrox commented on docs_src/src/pages/documentation/api_reference/openapi.mdx:
```diff
@@ -0,0 +1,203 @@
+    return r.query_params
+
+
+@app.delete("/users/:name", openapi_tags=["Name"])
```
could you add an endpoint for path params as well?

sansyrox commented on docs_src/src/pages/documentation/api_reference/openapi.mdx:
```diff
@@ -0,0 +1,203 @@
+async def delete(r: Request):
+    """Delete Name by ID"""
+    return r.query_params
+
```
same here

sansyrox commented on docs_src/src/pages/documentation/api_reference/openapi.mdx:
```diff
@@ -0,0 +1,203 @@
+@subrouter.delete("/:name")
+async def subrouter_delete(r: Request):
+    """Delete Name by ID"""
+    return r.query_params
```
path params here

sansyrox commented on docs_src/src/pages/documentation/api_reference/openapi.mdx:
```diff
@@ -0,0 +1,203 @@
+@subrouter.delete("/:name")
+async def subrouter_delete(r):
+    """Delete Name by ID"""
+    return jsonify(r.query_params)
```
path params here

sansyrox replied:
and no jsonify

sansyrox commented on docs_src/src/pages/documentation/api_reference/openapi.mdx:
```diff
@@ -0,0 +1,203 @@
+
+
+@subrouter.delete("/:name")
+async def subrouter_delete(r: Request):
```
add a response type in output please

VishnuSanal replied:
didn't get you?

sansyrox commented on docs_src/src/pages/documentation/api_reference/openapi.mdx:
```diff
@@ -0,0 +1,203 @@
+
+<div className="not-prose">
+  <Button
+    href="/documentation/example_app/templates"
```
this is incorrect @VishnuSanal . please have a look at this

sansyrox commented on docs_src/src/pages/documentation/example_app/openapi.mdx:
```diff
@@ -0,0 +1,203 @@
+
+
+@app.delete("/users/:name", openapi_tags=["Name"])
+async def delete(r):
```
where are the query params here?

sansyrox commented on docs_src/src/pages/documentation/example_app/openapi.mdx:
```diff
@@ -0,0 +1,203 @@
+
+
+@app.delete("/users/:name", openapi_tags=["Name"])
+async def delete(r: Request):
```
no query params?

sansyrox commented on docs_src/src/pages/documentation/example_app/openapi.mdx:
```diff
@@ -0,0 +1,203 @@
+@subrouter.get("/:name")
+async def subrouter_get(r):
+    """Get Name by ID"""
+    return jsonify(r.query_params)
```
shouldn't use jsonify.

sansyrox commented on docs_src/src/pages/documentation/example_app/openapi.mdx:
```diff
@@ -0,0 +1,203 @@
+@subrouter.delete("/:name")
+async def subrouter_delete(r):
+    """Delete Name by ID"""
+    return jsonify(r.query_params)
```
jsonify

sansyrox commented on integration_tests/base_routes.py:
```diff
@@ -832,6 +833,34 @@ def sync_router_di(request, router_dependencies):
+        title="Sample Pet Store App",
+        description=" This is a sample server for a pet store.",
+        termsOfService=" https://example.com/terms/",
+        version=" 1.0.1",
```
```suggestion
        version="1.0.1",
```

sansyrox commented on integration_tests/base_routes.py:
```diff
@@ -832,6 +833,34 @@ def sync_router_di(request, router_dependencies):
+    info=OpenAPIInfo(
+        title="Sample Pet Store App",
+        description=" This is a sample server for a pet store.",
+        termsOfService=" https://example.com/terms/",
```
```suggestion
        termsOfService="https://example.com/terms/",
```

sansyrox commented on integration_tests/base_routes.py:
```diff
@@ -832,6 +833,34 @@ def sync_router_di(request, router_dependencies):
+            email="support@example.com",
+        ),
+        license=License(
+            name="Apache 2.0",
```
why do we need a different licence here?

sansyrox commented on robyn/__init__.py:
```diff
@@ -302,110 +310,193 @@ def inner(handler):
+        endpoint: str,
+        const: bool = False,
+        auth_required: bool = False,
+        openapi_tags: list = ["default"],
```
this should default to 
```suggestion
        openapi_tags: list = ["get"],
```

sansyrox commented on robyn/__init__.py:
```diff
@@ -302,110 +310,193 @@ def inner(handler):
+        self,
+        endpoint: str,
+        auth_required: bool = False,
+        openapi_tags: list = ["default"],
```
and so on

sansyrox commented on robyn/__init__.py:
```diff
@@ -440,35 +533,35 @@ def configure_authentication(self, authentication_handler: AuthenticationHandler
 
 class SubRouter(Robyn):
     def __init__(self, file_object: str, prefix: str = "", config: Config = Config()) -> None:
```
```suggestion
    def __init__(self, file_object: str, prefix: str = "", config: Config = Config(), openapi = OpenAPI()) -> None:
```

This is to avoid coupling.

sansyrox commented on robyn/__init__.py:
```diff
@@ -440,35 +533,35 @@ def configure_authentication(self, authentication_handler: AuthenticationHandler
 class SubRouter(Robyn):
     def __init__(self, file_object: str, prefix: str = "", config: Config = Config()) -> None:
-        super().__init__(file_object, config)
+        super().__init__(file_object, config, OpenAPI())
```
use like config in the constructor

sansyrox commented on robyn/__init__.py:
```diff
@@ -440,35 +533,35 @@ def configure_authentication(self, authentication_handler: AuthenticationHandler
 
-    def get(self, endpoint: str, const: bool = False):
-        return super().get(self.__add_prefix(endpoint), const)
+    def get(self, endpoint: str, const: bool = False, openapi_tags: list = ["default"]):
```
same here. need to change the default tags

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,275 @@
+                    }
+                )
+                endpoint_with_path_params_wrapped_in_braces += "/{" + path_param_name + "}"
+
+        if query_params:
+            for query_param in query_params.__annotations__:
+                # ugly hack!
+                # returns "<class 'int'>" -- this line strips out the type a.k.a int from it
+                query_param_type = re.findall("'[a-z]+'", str(query_params.__annotations__[query_param]))[0].replace("'", "")
```
still need to have a look at this

VishnuSanal replied:
hey @sansyrox I can't figure out a way to do it, can you give some help & point me in the right direction?!

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,275 @@
+from pathlib import Path
+from typing import Dict, List, Optional, TypedDict
+
+import robyn
```
don't do this. this can induce circular imports. 

Import the module using absolute paths

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,275 @@
+        """
+        self.dump_openapi_spec_file()
+        html_file = str(Path("./robyn/swagger.html"))
+        return robyn.serve_html(html_file)
```
```suggestion
        return serve_html(html_file)
```

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,275 @@
+        side effect: this function also dumps the openapi json file
+        @return: the swagger html page
+        """
+        self.dump_openapi_spec_file()
```
this doesn't do anything, just returns a string. Where are you dumping the file?

sansyrox replied:
🔴

VishnuSanal replied:
I have updated the implementation. I was assuming something else when I fisr wrote the code, my mistake!

sansyrox reviewed (commented):
Hey @VishnuSanal 👋 

I have some follow ups

VishnuSanal commented:
@sansyrox I have addressed the review comments. PTAL. also, I have added the example app back. will remove it later. I feel that let's fix the errors/issues in the example application (to be added to the docs) & then add it to the docs -- than changing the same content on multiple places (api ref & example app && typoed & untyped)!

sansyrox commented on docs_src/src/components/documentation/Navigation.jsx:
```diff
@@ -210,6 +210,10 @@ export const navigation = [
       },
       { title: 'Deployment', href: '/documentation/example_app/deployment' },
+      {
+        title: 'OpenAPI Specification',
```
```suggestion
        title: 'OpenAPI Documentation',
```

sansyrox commented on docs_src/src/components/documentation/Guides.jsx:
```diff
@@ -34,6 +34,12 @@ const guides = [
   },
+  {
+    href: '/documentation/example_app/openapi',
+    name: 'OpenAPI Specification',
```
```suggestion
    name: 'OpenAPI Documentation',
```

sansyrox commented on docs_src/src/components/documentation/ApiDocs.jsx:
```diff
@@ -106,6 +106,12 @@ const guides = [
   },
+  {
+    href: '/documentation/api_reference/openapi',
+    name: 'OpenAPI Specification',
```
```suggestion
    name: 'OpenAPI Documentation',
```

sansyrox commented on integration_tests/base_routes.py:
```diff
@@ -832,6 +829,34 @@ def sync_router_di(request, router_dependencies):
 
 
+# ===== OpenAPI =====
+app.openapi = OpenAPI(
```
this is optional, right?

VishnuSanal replied:
yep.

sansyrox commented on integration_tests/test_openapi.py:
```diff
@@ -0,0 +1,58 @@
+
+@pytest.mark.benchmark
+def test_add_subrouter_paths():
+    openapi_spec = ast.literal_eval(get("/openapi.json").text)
```
what? no. please use a dictionary directly instead of using the ast

sansyrox commented on robyn/__init__.py:
```diff
@@ -302,110 +310,193 @@ def inner(handler):
         """
 
         def inner(handler):
+            self.openapi.add_openapi_path_obj("get", endpoint, inspect.getdoc(handler), openapi_tags, inspect.signature(handler))
```
one question what happens if there is no docstring? i.e in `inspect.getdoc`?

VishnuSanal replied:
prints "None" -- I will add a fallback value.

sansyrox replied:
@VishnuSanal , this is unresolved right?

VishnuSanal replied:
no, this is fixed. the summary falls back to "No summary provided".

sansyrox commented on robyn/__init__.py:
```diff
@@ -302,110 +310,193 @@ def inner(handler):
-        :param endpoint str: endpoint to server the route
+        :param endpoint str: endpoint for the route added
+        :param auth_required bool: represents if the route needs authentication or not
+        :param openapi_tags: tags -- for grouping of endpoints in the openapi spec
```
is the type `tags` ?

sansyrox commented on robyn/__init__.py:
```diff
@@ -302,110 +310,193 @@ def inner(handler):
-        :param endpoint str: endpoint to server the route
+        :param endpoint str: endpoint for the route added
+        :param auth_required bool: represents if the route needs authentication or not
+        :param openapi_tags: tags -- for grouping of endpoints in the openapi spec
```
same for all the types. this should not be tags

sansyrox commented on robyn/__init__.py:
```diff
@@ -439,36 +532,36 @@ def configure_authentication(self, authentication_handler: AuthenticationHandler
-    def __init__(self, file_object: str, prefix: str = "", config: Config = Config()) -> None:
-        super().__init__(file_object, config)
+    def __init__(self, file_object: str, prefix: str = "", config: Config = Config(), openapi: OpenAPI = OpenAPI()) -> None:
+        super().__init__(file_object, config, OpenAPI())
```
```suggestion
        super().__init__(file_object, config, openapi)
```

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,264 @@
+from pathlib import Path
+from typing import Dict, List, Optional, TypedDict
+
+from .responses import FileResponse, serve_html
```
should use absolute paths.

```suggestion
from robyn.responses import FileResponse, serve_html
```

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,264 @@
+from dataclasses import asdict, dataclass, field
+from inspect import Signature
+from pathlib import Path
+from typing import Dict, List, Optional, TypedDict
```
```suggestion
from typing import Optional, TypedDict
```

We can just use `dict` and `list`

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,264 @@
+    termsOfService: Optional[str] = None
+    contact: Contact = field(default_factory=Contact)
+    license: License = field(default_factory=License)
+    servers: List[Server] = field(default_factory=list)
```
```suggestion
    servers: list[Server] = field(default_factory=list)
```

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,264 @@
+
+    def json_handler(self) -> str:
+        """
+        Handler to the openapi spec json object to be deployed to the endpoint `/openapi.json`
```
`json_handler` would usually return a json object. we should name it something else.

sansyrox commented on robyn/__init__.py:
```diff
@@ -302,110 +310,193 @@ def inner(handler):
+        self,
+        endpoint: str,
+        auth_required: bool = False,
+        openapi_tags: list = ["post"],
```
```suggestion
        openapi_tags: list[str] = ["post"],
```
No?

sansyrox replied:
@VishnuSanal , this should be done across all the types.

sansyrox commented on robyn/__init__.py:
```diff
@@ -302,110 +310,193 @@ def inner(handler):
+        endpoint: str,
+        const: bool = False,
+        auth_required: bool = False,
+        openapi_tags: list = ["get"],
```
same here `list[str]`

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,264 @@
+                endpoint_with_path_params_wrapped_in_braces += "/{" + path_param_name + "}"
+
+        if query_params:
+            for query_param in query_params.__annotations__:
```
You can use `get_type_hints` form `typing`.

I used AI for some suggestion as I not at home atm. But this sample can come in hand

```
from typing import get_type_hints, TypedDict, Union

def get_path_obj(self, endpoint: str, summary: str, tags: list, query_params: TypedDict = None) -> (str, dict):
    """
    Get the "path" openapi object according to spec

    :param endpoint: the endpoint to be added
    :param summary: short summary of the endpoint (to be fetched from the endpoint defenition by default)
    :param tags: tags -- for grouping of endpoints
    :param query_params: query params for the function

    :return: a tuple containing the endpoint with path params wrapped in braces and the "path" openapi object
    according to spec
    """
    # robyn has paths like /:url/:etc whereas openapi requires path like /{url}/{path}
    # this function is used for converting path params to the required form
    # initialized with endpoint for handling endpoints without path params
    endpoint_with_path_params_wrapped_in_braces = endpoint

    endpoint_path_params_split = endpoint.split(":")
    openapi_parameter_object = []

    if len(endpoint_path_params_split) > 1:
        endpoint_without_path_params = endpoint_path_params_split[0]

        endpoint_with_path_params_wrapped_in_braces = (
            endpoint_without_path_params[:-1] if endpoint_without_path_params.endswith("/") else endpoint_without_path_params
        )

        for path_param in endpoint_path_params_split[1:]:
            path_param_name = path_param[:-1] if path_param.endswith("/") else path_param

            openapi_parameter_object.append(
                {
                    "name": path_param_name,
                    "in": "path",
                    "required": True,
                    "schema": {"type": "string"},
                }
            )
            endpoint_with_path_params_wrapped_in_braces += "/{" + path_param_name + "}"

    if query_params:
        # Use get_type_hints to retrieve the actual types from the annotations
        query_param_types = get_type_hints(query_params)
        
        for query_param, query_param_type in query_param_types.items():
            openapi_type, custom_type = self.get_openapi_type(query_param_type)

            param_schema = {"type": openapi_type}
            if custom_type:
                param_schema["$ref"] = f"#/components/schemas/{custom_type}"

            openapi_parameter_object.append(
                {
                    "name": query_param,
                    "in": "query",
                    "required": False,
                    "schema": param_schema,
                }
            )

    return endpoint_with_path_params_wrapped_in_braces, {
        "summary": summary,
        "tags": tags,
        "parameters": openapi_parameter_object,
        "responses": {
            "200": {
                "description": "Successful Response",
            }
        },
    }

def get_openapi_type(self, python_type: Union[type, str]) -> (str, Optional[str]):
    """
    Converts Python types to OpenAPI types, handling custom types.

    :param python_type: The Python type to be converted
    :return: A tuple with OpenAPI type as a string and custom type name if applicable
    """
    type_mapping = {
        int: "integer",
        str: "string",
        bool: "boolean",
        float: "number",
        dict: "object",
        list: "array",
        # Add more mappings as necessary
    }

    # If it's a standard type, return the OpenAPI type directly
    if python_type in type_mapping:
        return type_mapping[python_type], None

    # If it's a custom type (likely a class), return its name for the $ref
    if hasattr(python_type, '__name__'):
        return "object", python_type.__name__

    # Handle cases where python_type is provided as a string
    if isinstance(python_type, str):
        python_type = eval(python_type)
        if python_type in type_mapping:
            return type_mapping[python_type], None
        elif hasattr(python_type, '__name__'):
            return "object", python_type.__name__

    # Default to "string" if type is not found
    return "string", None

```

sansyrox commented on testing_application/app.py:
```diff
@@ -0,0 +1,82 @@
+
+
+@subrouter.delete("/:name")
+async def subrouter_delete(r: Request):
```
```suggestion
async def subrouter_delete(r: Request, path_params: Type):
```
??

VishnuSanal replied:
what??! path params are defaulted to strings as we discussed previously!

sansyrox replied:
Could you please mention this fact in the docs?

VishnuSanal replied:
alirght

sansyrox reviewed (commented):
Hey @VishnuSanal 👋 

I have some comments.

sansyrox commented on integration_tests/base_routes.py:
```diff
@@ -3,6 +3,10 @@
 from collections import defaultdict
 from typing import Optional
 
+from robyn.robyn import Headers
```
```suggestion
from robyn import Headers
```

This can be imported like this

sansyrox commented on robyn/__init__.py:
```diff
@@ -302,110 +310,193 @@ def inner(handler):
+        endpoint: str,
+        const: bool = False,
+        auth_required: bool = False,
+        openapi_tags: List[str] = ["get"],
```
```suggestion
        openapi_tags: list[str] = ["get"],
```

sansyrox commented on robyn/__init__.py:
```diff
@@ -302,110 +310,193 @@ def inner(handler):
+        :param endpoint str: endpoint for the route added
+        :param const bool: represents if the handler is a const function or not
+        :param auth_required bool: represents if the route needs authentication or not
+        :param openapi_tags: List[str] -- for grouping of endpoints in the openapi spec
```
```suggestion
        :param openapi_tags: list[str] -- for grouping of endpoints in the openapi spec
```

This should be the types for the entire file

sansyrox commented on robyn/__init__.py:
```diff
@@ -302,110 +310,193 @@ def inner(handler):
+        self,
+        endpoint: str,
+        auth_required: bool = False,
+        openapi_tags: List[str] = ["post"],
```
```suggestion
        openapi_tags: list[str] = ["post"],
```

sansyrox commented on robyn/__init__.py:
```diff
@@ -439,36 +532,36 @@ def configure_authentication(self, authentication_handler: AuthenticationHandler
 
-    def get(self, endpoint: str, const: bool = False):
-        return super().get(self.__add_prefix(endpoint), const)
+    def get(self, endpoint: str, const: bool = False, openapi_tags: List[str] = ["get"]):
```
```suggestion
    def get(self, endpoint: str, const: bool = False, openapi_tags: list[str] = ["get"]):
```

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,287 @@
+        for path in paths:
+            self.openapi_spec["paths"][path] = paths[path]
+
+    def get_path_obj(self, endpoint: str, summary: str, tags: List[str], query_params: TypedDict = None) -> (str, dict):
```
```suggestion
    def get_path_obj(self, endpoint: str, summary: str, tags: List[str], query_params: Optional[TypedDict] = None) -> (str, dict):
```

?

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,287 @@
+        :param endpoint: str the endpoint to be added
+        :param summary: Optional[str] short summary of the endpoint (to be fetched from the endpoint defenition by default)
+        :param tags: List[str] for grouping of endpoints
+        :param query_params: TypedDict query params for the function
```
```suggestion
        :param query_params: Optional[TypedDict] query params for the function
```

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,287 @@
+            self.openapi_spec["paths"][modified_endpoint] = {}
+        self.openapi_spec["paths"][modified_endpoint][route_type] = path_obj
+
+    def add_subrouter_paths(self, subrouter_openapi):
```
```suggestion
    def add_subrouter_paths(self, subrouter_openapi):
```

Could you add a type here?

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,287 @@
+            "externalDocs": asdict(self.info.externalDocs) if self.info.externalDocs.url else None,
+        }
+
+    def add_openapi_path_obj(self, route_type: str, endpoint: str, openapi_summary: str, openapi_tags: list, signature: Signature = None):
```
```suggestion
    def add_openapi_path_obj(self, route_type: str, endpoint: str, openapi_summary: str, openapi_tags: list, signature: Optional[Signature] = None):
```

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,287 @@
+        """
+        Adds the given path to openapi spec
+
+        :param route_type: the http method as string (get, post ...)
+        :param endpoint: the endpoint to be added
+        :param openapi_summary: short summary of the endpoint (to be fetched from the endpoint defenition by default)
+        :param openapi_tags: tags -- for grouping of endpoints
+        :param signature: the function signature -- to grab the typed dict annotations for query params
```
Please add types here

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,287 @@
+        :param tags: List[str] for grouping of endpoints
+        :param query_params: TypedDict query params for the function
+
+        :return: a tuple containing the endpoint with path params wrapped in braces and the "path" openapi object
```
needs a return type

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,287 @@
+    def get_openapi_docs_page(self) -> FileResponse:
+        """
+        Handler to the swagger html page to be deployed to the endpoint `/docs`
+        @return: the swagger html page
```
return type missing

sansyrox commented on robyn/openapi.py:
```diff
@@ -0,0 +1,287 @@
+    def get_openapi_config_string(self) -> str:
+        """
+        Handler to the openapi spec json object to be deployed to the endpoint `/openapi.json`
+        @return: a JSON string representing the openapi spec
```
same

sansyrox commented on testing_application/app.py:
```diff
@@ -0,0 +1,82 @@
+    openapi=OpenAPI(
+        info=OpenAPIInfo(
+            title="Sample App",
+            description=" This is a sample server application.",
```
```suggestion
            description="This is a sample server application.",
```

sansyrox commented on testing_application/app.py:
```diff
@@ -0,0 +1,82 @@
+        info=OpenAPIInfo(
+            title="Sample App",
+            description=" This is a sample server application.",
+            termsOfService=" https://example.com/terms/",
```
```suggestion
            termsOfService="https://example.com/terms/",
```

sansyrox commented on testing_application/app.py:
```diff
@@ -0,0 +1,82 @@
+            title="Sample App",
+            description=" This is a sample server application.",
+            termsOfService=" https://example.com/terms/",
+            version=" 1.2.2",
```
```suggestion
            version="1.2.2",
```

sansyrox commented on testing_application/app.py:
```diff
@@ -0,0 +1,82 @@
+    year: int
+
+
+@app.get("/api/v1/:name", openapi_tags=["Name"])
+async def get(r: Request, query_params=GetRequestParams):
+    """Get Name by ID"""
+    return r.query_params
```
where are the query params in this route?

VishnuSanal replied:
what?

sansyrox replied:
@VishnuSanal `:/name` is a path parameter. You are returning query parameter.

sansyrox replied:
@VishnuSanal , where is the path param being used?

VishnuSanal replied:
> @VishnuSanal , where is the path param being used?

um... it is not "used", but that will show up in the swagger docs as a field.

![image](https://github.com/user-attachments/assets/d9efea34-75b7-40df-855c-5cf226fd51cb)

should I remove it?

sansyrox replied:
@VishnuSanal , yes please. Let us keep the test routes specific to their function.

VishnuSanal replied:
@sansyrox done, PTAL. :)

sansyrox commented on testing_application/app.py:
```diff
@@ -0,0 +1,82 @@
+            title="Sample App",
+            description="This is a sample server application.",
+            termsOfService="https://example.com/terms/",
+            version="1.2.2",
```
why do we have random version names here?

VishnuSanal replied:
ill change to v1.0.0~

sansyrox commented on docs_src/src/pages/documentation/api_reference/openapi.mdx:
````diff
@@ -0,0 +1,224 @@
+
+<CodeGroup title="Basic App">
+
+```python {{ title: 'untyped' }}
+from typing import TypedDict
+
+from robyn import Robyn, OpenAPI
+from robyn.openapi import OpenAPIInfo, Contact, License, ExternalDocumentation, Components
+
````
this is an untyped example

sansyrox reviewed (approved):
LGTM! Great work @VishnuSanal

sansyrox reviewed (approved):
LGTM! Great work @VishnuSanal 👍

tavallaie commented:
Thansk @VishnuSanal
